// evergreen-ui/esm/radio/src/RadioGroup.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Pane } from '../../layers';
import Radio from './Radio';

var RadioGroup = function (_PureComponent) {
  _inherits(RadioGroup, _PureComponent);

  function RadioGroup(props, context) {
    _classCallCheck(this, RadioGroup);

    var _this = _possibleConstructorReturn(this, (RadioGroup.__proto__ || Object.getPrototypeOf(RadioGroup)).call(this, props, context));

    _this.handleChange = function (value) {
      _this.setState({
        selected: value
      });
      _this.props.onChange(value);
    };

    _this.state = {
      selected: (props.items.find(function (item) {
        return item.isDefaultSelected;
      }) || {}).value
    };
    return _this;
  }

  _createClass(RadioGroup, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          label = _props.label,
          items = _props.items,
          onChange = _props.onChange,
          props = _objectWithoutProperties(_props, ['label', 'items', 'onChange']);

      var selected = this.state.selected;


      return React.createElement(
        Pane,
        _extends({ role: 'group', 'aria-label': label }, props),
        items.map(function (item) {
          return React.createElement(Radio, {
            key: item.value,
            name: item.name,
            value: item.value,
            label: item.label,
            checked: selected === item.value,
            disabled: item.isDisabled,
            onChange: function onChange() {
              return _this2.handleChange(item.value);
            }
          });
        })
      );
    }
  }]);

  return RadioGroup;
}(PureComponent);

RadioGroup.propTypes = _extends({}, Pane.propTypes, {
  label: PropTypes.string,
  onChange: PropTypes.func,
  items: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
    label: PropTypes.node,
    isDisabled: PropTypes.bool,
    isDefaultSelected: PropTypes.bool
  })).isRequired
});
RadioGroup.defaultProps = {
  items: []
};
export default RadioGroup;
