// evergreen-ui/commonjs/side-sheet/src/SideSheet.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uiBox = require('ui-box');

var _layers = require('../../layers');

var _overlay = require('../../overlay');

var _SheetClose = require('./SheetClose');

var _SheetClose2 = _interopRequireDefault(_SheetClose);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var paneProps = {
  height: '100vh',
  position: 'absolute',
  right: 0
};

var animationEasing = {
  deceleration: 'cubic-bezier(0.0, 0.0, 0.2, 1)',
  acceleration: 'cubic-bezier(0.4, 0.0, 1, 1)'
};

var ANIMATION_DURATION = 240;

var slideInAnimation = _uiBox.css.keyframes('slideInAnimation', {
  from: {
    transform: 'translateX(100%)'
  },
  to: {
    transform: 'translateX(0)'
  }
});

var slideOutAnimation = _uiBox.css.keyframes('slideOutAnimation', {
  from: {
    transform: 'translateX(0)'
  },
  to: {
    transform: 'translateX(100%)'
  }
});

var animationStyles = {
  transform: 'translateX(100%)',
  '&[data-state="entering"], &[data-state="entered"]': {
    animation: slideInAnimation + ' ' + ANIMATION_DURATION + 'ms ' + animationEasing.deceleration + ' both'
  },
  '&[data-state="exiting"]': {
    animation: slideOutAnimation + ' ' + ANIMATION_DURATION + 'ms ' + animationEasing.acceleration + ' both'
  }
};

var SideSheet = function (_React$Component) {
  _inherits(SideSheet, _React$Component);

  function SideSheet() {
    _classCallCheck(this, SideSheet);

    return _possibleConstructorReturn(this, (SideSheet.__proto__ || Object.getPrototypeOf(SideSheet)).apply(this, arguments));
  }

  _createClass(SideSheet, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          width = _props.width,
          containerProps = _props.containerProps,
          props = _objectWithoutProperties(_props, ['children', 'width', 'containerProps']);

      return _react2.default.createElement(
        _overlay.Overlay,
        props,
        function (_ref) {
          var state = _ref.state,
              close = _ref.close;
          return _react2.default.createElement(
            _layers.Pane,
            _extends({
              width: width
            }, paneProps, {
              css: animationStyles,
              'data-state': state
            }),
            _react2.default.createElement(_SheetClose2.default, { 'data-state': state, isClosing: false, onClick: close }),
            _react2.default.createElement(
              _layers.Pane,
              _extends({
                elevation: 4,
                backgroundColor: 'white',
                overflowY: 'auto',
                'data-state': state,
                width: width
              }, paneProps, containerProps),
              typeof children === 'function' ? children({ state: state, close: close }) : children
            )
          );
        }
      );
    }
  }]);

  return SideSheet;
}(_react2.default.Component);

SideSheet.propTypes = _extends({}, _overlay.Overlay.propTypes, {

  /**
   * Width of the SideSheet.
   */
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),

  /**
   * Properties to pass through the SideSheet container Pane.
   */
  containerProps: _propTypes2.default.object
});
SideSheet.defaultProps = {
  width: 620
};
exports.default = SideSheet;
