// evergreen-ui/esm/segmented-control/src/SegmentedControl.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Box from 'ui-box';
import SegmentedControlAppearances from './styles/SegmentedControlAppearances';
import SegmentedControlRadio from './SegmentedControlRadio';

var keysSegmentedControlAppearances = Object.keys(SegmentedControlAppearances);
var radioCount = 1; // Used for generating unique input names

var SegmentedControl = function (_PureComponent) {
  _inherits(SegmentedControl, _PureComponent);

  function SegmentedControl(props, context) {
    _classCallCheck(this, SegmentedControl);

    var _this = _possibleConstructorReturn(this, (SegmentedControl.__proto__ || Object.getPrototypeOf(SegmentedControl)).call(this, props, context));

    _this.handleChange = function (value) {
      // Save a render cycle when it's a controlled input
      if (!_this.props.value) {
        _this.setState({ value: value });
      }

      if (_this.props.onChange) {
        _this.props.onChange(value);
      }
    };

    _this.state = {
      value: props.defaultValue || props.options[0].value
    };

    _this.name = 'SegmentedControl-' + radioCount;
    radioCount += 1;
    return _this;
  }

  _createClass(SegmentedControl, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          name = _props.name,
          height = _props.height,
          options = _props.options,
          onChange = _props.onChange,
          appearance = _props.appearance,
          defaultValue = _props.defaultValue,
          props = _objectWithoutProperties(_props, ['name', 'height', 'options', 'onChange', 'appearance', 'defaultValue']);

      // Allows it to behave like a controlled input


      var value = this.props.value || this.state.value;
      return React.createElement(
        Box,
        _extends({ display: 'flex', marginRight: -1, height: height }, props),
        options.map(function (option, index) {
          return React.createElement(SegmentedControlRadio, {
            key: option.value,
            name: name,
            label: option.label,
            value: option.value,
            height: height,
            checked: value === option.value,
            onChange: _this2.handleChange,
            appearance: appearance,
            isFirstItem: index === 0,
            isLastItem: index === options.length - 1
          });
        })
      );
    }
  }]);

  return SegmentedControl;
}(PureComponent);

SegmentedControl.propTypes = _extends({}, Box.propTypes, {
  appearance: PropTypes.oneOf(keysSegmentedControlAppearances),
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.node.isRequired,
    value: PropTypes.string.isRequired
  })).isRequired,
  name: PropTypes.string,
  value: PropTypes.string,
  height: PropTypes.number,
  onChange: PropTypes.func,
  defaultValue: PropTypes.string
});
SegmentedControl.defaultProps = {
  height: 32,
  appearance: 'default'
};
export default SegmentedControl;
