// evergreen-ui/commonjs/tooltip/src/TooltipStateless.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uiBox = require('ui-box');

var _uiBox2 = _interopRequireDefault(_uiBox);

var _colors = require('../../colors');

var _typography = require('../../typography');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TooltipStateless = function (_PureComponent) {
  _inherits(TooltipStateless, _PureComponent);

  function TooltipStateless() {
    _classCallCheck(this, TooltipStateless);

    return _possibleConstructorReturn(this, (TooltipStateless.__proto__ || Object.getPrototypeOf(TooltipStateless)).apply(this, arguments));
  }

  _createClass(TooltipStateless, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['children']);

      var child = void 0;
      if (typeof children === 'string') {
        child = _react2.default.createElement(
          _typography.Text,
          { display: 'block', lineHeight: '1.25em', color: 'white', size: 400 },
          children
        );
      } else {
        child = children;
      }
      return _react2.default.createElement(
        _uiBox2.default,
        _extends({}, TooltipStateless.styles, props),
        child
      );
    }
  }]);

  return TooltipStateless;
}(_react.PureComponent);

TooltipStateless.propTypes = {
  children: _propTypes2.default.node
};
TooltipStateless.styles = {
  backgroundColor: _colors.colors.neutral['400A'],
  borderRadius: 3,
  paddingX: 8,
  paddingY: 4,
  maxWidth: 240
};
exports.default = TooltipStateless;
