// evergreen-ui/esm/select-menu/src/SelectMenu.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import arrify from 'arrify';
import { Popover } from '../../popover';
import SelectMenuContent from './SelectMenuContent';
import OptionShapePropType from './OptionShapePropType';
import SelectedPropType from './SelectedPropType';

var SelectMenu = function (_PureComponent) {
  _inherits(SelectMenu, _PureComponent);

  function SelectMenu(props, context) {
    _classCallCheck(this, SelectMenu);

    var _this = _possibleConstructorReturn(this, (SelectMenu.__proto__ || Object.getPrototypeOf(SelectMenu)).call(this, props, context));

    _this.state = {};
    return _this;
  }

  _createClass(SelectMenu, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          title = _props.title,
          width = _props.width,
          height = _props.height,
          options = _props.options,
          selected = _props.selected,
          hasTitle = _props.hasTitle,
          hasFilter = _props.hasFilter,
          props = _objectWithoutProperties(_props, ['title', 'width', 'height', 'options', 'selected', 'hasTitle', 'hasFilter']);

      return React.createElement(Popover, _extends({
        minWidth: width,
        minHeight: height,
        content: function content(_ref) {
          var close = _ref.close;
          return React.createElement(SelectMenuContent, {
            width: width,
            height: height,
            options: options,
            title: title,
            hasFilter: hasFilter,
            hasTitle: hasTitle,
            listProps: {
              onSelect: function onSelect(item) {
                _this2.props.onSelect(item);
              },
              selected: arrify(selected)
            },
            close: close
          });
        }
      }, props));
    }
  }]);

  return SelectMenu;
}(PureComponent);

SelectMenu.propTypes = {
  title: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  options: PropTypes.arrayOf(OptionShapePropType),
  children: PropTypes.node,
  onSelect: PropTypes.func,
  selected: SelectedPropType,
  hasTitle: PropTypes.bool,
  hasFilter: PropTypes.bool
};
SelectMenu.defaultProps = {
  onSelect: function onSelect() {},
  width: 240,
  height: 248
};
export default SelectMenu;
