// evergreen-ui/esm/tooltip/src/Tooltip.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Positioner } from '../../positioner';
import TooltipStateless from './TooltipStateless';

var Tooltip = function (_PureComponent) {
  _inherits(Tooltip, _PureComponent);

  function Tooltip(props, context) {
    _classCallCheck(this, Tooltip);

    var _this = _possibleConstructorReturn(this, (Tooltip.__proto__ || Object.getPrototypeOf(Tooltip)).call(this, props, context));

    _this.getTargetRect = function () {
      return _this.targetNode.getBoundingClientRect();
    };

    _this.getRef = function (ref) {
      _this.targetNode = ref;
    };

    _this.show = function () {
      if (_this.state.isShown) return;
      _this.setState({
        isShown: true,
        targetRect: _this.getTargetRect()
      });
    };

    _this.hide = function () {
      if (!_this.state.isShown) return;
      _this.setState({
        isShown: false,
        targetRect: _this.getTargetRect()
      });
    };

    _this.state = {
      isShown: false
    };
    return _this;
  }

  _createClass(Tooltip, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          isShown = _props.isShown,
          content = _props.content,
          children = _props.children,
          statelessProps = _props.statelessProps,
          props = _objectWithoutProperties(_props, ['isShown', 'content', 'children', 'statelessProps']);

      var _state = this.state,
          stateIsShown = _state.isShown,
          targetRect = _state.targetRect;


      var shown = isShown || stateIsShown;

      return [typeof children === 'function' ? children({
        targetRect: targetRect,
        show: this.show,
        hide: this.hide,
        getRef: this.getRef,
        isShown: shown,
        key: 'tooltip-child'
      }) : React.cloneElement(children, _extends({
        onMouseEnter: this.show,
        onMouseLeave: this.hide,
        innerRef: function innerRef(ref) {
          _this2.getRef(ref);
        }
      }, shown ? { 'data-tooltip-opened': true } : {}, {
        key: 'tooltip-child'
      })), React.createElement(
        Positioner,
        _extends({
          key: 'tooltip-positioner',
          targetRect: targetRect,
          isShown: shown,
          initialScale: 0.95,
          targetOffset: 4
        }, props),
        function (_ref) {
          var css = _ref.css,
              style = _ref.style,
              state = _ref.state,
              getRef = _ref.getRef;
          return React.createElement(
            TooltipStateless,
            _extends({
              innerRef: function innerRef(ref) {
                return getRef(ref);
              },
              'data-state': state,
              css: css,
              style: style
            }, statelessProps),
            content
          );
        }
      )];
    }
  }]);

  return Tooltip;
}(PureComponent);

Tooltip.propTypes = _extends({}, Positioner.propTypes, {
  content: PropTypes.node,
  isShown: PropTypes.bool,
  children: PropTypes.node,
  tooltipProps: PropTypes.object,
  statelessProps: PropTypes.objectOf(TooltipStateless.propTypes)
});
export default Tooltip;
