// evergreen-ui/esm/tooltip/src/TooltipStateless.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Box from 'ui-box';
import { colors } from '../../colors';
import { Text } from '../../typography';

var TooltipStateless = function (_PureComponent) {
  _inherits(TooltipStateless, _PureComponent);

  function TooltipStateless() {
    _classCallCheck(this, TooltipStateless);

    return _possibleConstructorReturn(this, (TooltipStateless.__proto__ || Object.getPrototypeOf(TooltipStateless)).apply(this, arguments));
  }

  _createClass(TooltipStateless, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['children']);

      var child = void 0;
      if (typeof children === 'string') {
        child = React.createElement(
          Text,
          { display: 'block', lineHeight: '1.25em', color: 'white', size: 400 },
          children
        );
      } else {
        child = children;
      }
      return React.createElement(
        Box,
        _extends({}, TooltipStateless.styles, props),
        child
      );
    }
  }]);

  return TooltipStateless;
}(PureComponent);

TooltipStateless.propTypes = {
  children: PropTypes.node
};
TooltipStateless.styles = {
  backgroundColor: colors.neutral['400A'],
  borderRadius: 3,
  paddingX: 8,
  paddingY: 4,
  maxWidth: 240
};
export default TooltipStateless;
