// evergreen-ui/commonjs/select-menu/src/SelectMenuContent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _buttons = require('../../buttons');

var _layers = require('../../layers');

var _typography = require('../../typography');

var _OptionsList = require('./OptionsList');

var _OptionsList2 = _interopRequireDefault(_OptionsList);

var _OptionShapePropType = require('./OptionShapePropType');

var _OptionShapePropType2 = _interopRequireDefault(_OptionShapePropType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SelectMenuContent = function (_PureComponent) {
  _inherits(SelectMenuContent, _PureComponent);

  function SelectMenuContent(props, context) {
    _classCallCheck(this, SelectMenuContent);

    var _this = _possibleConstructorReturn(this, (SelectMenuContent.__proto__ || Object.getPrototypeOf(SelectMenuContent)).call(this, props, context));

    _this.handleKeyUp = function (e) {
      if (e.keyCode === 38) {
        _this.handleArrowUp();
      }

      if (e.keyCode === 40) {
        _this.handleArrowDown();
      }

      if (e.keyCode === 13) {
        _this.handleEnter();
      }
    };

    _this.state = {};
    return _this;
  }

  _createClass(SelectMenuContent, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener('keyup', this.handleKeyUp);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('keyup', this.handleKeyUp);
    }
  }, {
    key: 'getCurrentIndex',
    value: function getCurrentIndex() {
      var _props = this.props,
          options = _props.options,
          listProps = _props.listProps;
      var selected = listProps.selected;


      return options.findIndex(function (option) {
        return option.value === selected[0];
      });
    }
  }, {
    key: 'handleArrowUp',
    value: function handleArrowUp() {
      var _props2 = this.props,
          options = _props2.options,
          listProps = _props2.listProps;
      var onSelect = listProps.onSelect;


      var nextIndex = this.getCurrentIndex() - 1;

      if (nextIndex < 0) {
        nextIndex = options.length - 1;
      }

      onSelect(options[nextIndex]);
    }
  }, {
    key: 'handleArrowDown',
    value: function handleArrowDown() {
      var _props3 = this.props,
          options = _props3.options,
          listProps = _props3.listProps;
      var onSelect = listProps.onSelect;


      var nextIndex = this.getCurrentIndex() + 1;

      if (nextIndex === options.length) {
        nextIndex = 0;
      }

      onSelect(options[nextIndex]);
    }
  }, {
    key: 'handleEnter',
    value: function handleEnter() {
      var isSelected = this.getCurrentIndex() === -1;

      if (isSelected) {
        this.props.close();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props4 = this.props,
          width = _props4.width,
          height = _props4.height,
          options = _props4.options,
          hasTitle = _props4.hasTitle,
          hasFilter = _props4.hasFilter,
          close = _props4.close,
          title = _props4.title,
          listProps = _props4.listProps;


      var headerHeight = 40;
      var optionsListHeight = hasTitle ? height - headerHeight : height;

      return _react2.default.createElement(
        _layers.Pane,
        { width: width, height: height, display: 'flex', flexDirection: 'column' },
        hasTitle && _react2.default.createElement(
          _layers.Pane,
          {
            display: 'flex',
            alignItems: 'center',
            borderBottom: 'muted',
            padding: 8,
            height: headerHeight,
            boxSizing: 'border-box'
          },
          _react2.default.createElement(
            _layers.Pane,
            { flex: '1' },
            _react2.default.createElement(
              _typography.Heading,
              { size: 400 },
              title
            )
          ),
          _react2.default.createElement(_buttons.IconButton, {
            icon: 'close',
            appearance: 'ghost',
            height: 24,
            onClick: close
          })
        ),
        _react2.default.createElement(_OptionsList2.default, _extends({
          height: optionsListHeight,
          hasFilter: hasFilter,
          options: options
        }, listProps))
      );
    }
  }]);

  return SelectMenuContent;
}(_react.PureComponent);

SelectMenuContent.propTypes = {
  close: _propTypes2.default.func,
  title: _propTypes2.default.string,
  width: _propTypes2.default.number,
  height: _propTypes2.default.number,
  options: _propTypes2.default.arrayOf(_OptionShapePropType2.default),
  hasTitle: _propTypes2.default.bool,
  hasFilter: _propTypes2.default.bool,
  listProps: _propTypes2.default.shape(_OptionsList2.default.propTypes)
};
SelectMenuContent.defaultProps = {
  options: [],
  hasTitle: true,
  hasFilter: true
};
exports.default = SelectMenuContent;
