// evergreen-ui/commonjs/select-menu/src/SelectMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _arrify = require('arrify');

var _arrify2 = _interopRequireDefault(_arrify);

var _popover = require('../../popover');

var _SelectMenuContent = require('./SelectMenuContent');

var _SelectMenuContent2 = _interopRequireDefault(_SelectMenuContent);

var _OptionShapePropType = require('./OptionShapePropType');

var _OptionShapePropType2 = _interopRequireDefault(_OptionShapePropType);

var _SelectedPropType = require('./SelectedPropType');

var _SelectedPropType2 = _interopRequireDefault(_SelectedPropType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SelectMenu = function (_PureComponent) {
  _inherits(SelectMenu, _PureComponent);

  function SelectMenu(props, context) {
    _classCallCheck(this, SelectMenu);

    var _this = _possibleConstructorReturn(this, (SelectMenu.__proto__ || Object.getPrototypeOf(SelectMenu)).call(this, props, context));

    _this.state = {};
    return _this;
  }

  _createClass(SelectMenu, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          title = _props.title,
          width = _props.width,
          height = _props.height,
          options = _props.options,
          selected = _props.selected,
          hasTitle = _props.hasTitle,
          hasFilter = _props.hasFilter,
          props = _objectWithoutProperties(_props, ['title', 'width', 'height', 'options', 'selected', 'hasTitle', 'hasFilter']);

      return _react2.default.createElement(_popover.Popover, _extends({
        minWidth: width,
        minHeight: height,
        content: function content(_ref) {
          var close = _ref.close;
          return _react2.default.createElement(_SelectMenuContent2.default, {
            width: width,
            height: height,
            options: options,
            title: title,
            hasFilter: hasFilter,
            hasTitle: hasTitle,
            listProps: {
              onSelect: function onSelect(item) {
                _this2.props.onSelect(item);
              },
              selected: (0, _arrify2.default)(selected)
            },
            close: close
          });
        }
      }, props));
    }
  }]);

  return SelectMenu;
}(_react.PureComponent);

SelectMenu.propTypes = {
  title: _propTypes2.default.string,
  width: _propTypes2.default.number,
  height: _propTypes2.default.number,
  options: _propTypes2.default.arrayOf(_OptionShapePropType2.default),
  children: _propTypes2.default.node,
  onSelect: _propTypes2.default.func,
  selected: _SelectedPropType2.default,
  hasTitle: _propTypes2.default.bool,
  hasFilter: _propTypes2.default.bool
};
SelectMenu.defaultProps = {
  onSelect: function onSelect() {},
  width: 240,
  height: 248
};
exports.default = SelectMenu;
