// evergreen-ui/commonjs/corner-dialog/src/CornerDialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uiBox = require('ui-box');

var _Transition = require('react-transition-group/Transition');

var _Transition2 = _interopRequireDefault(_Transition);

var _layers = require('../../layers');

var _portal = require('../../portal');

var _typography = require('../../typography');

var _buttons = require('../../buttons');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var animationEasing = {
  deceleration: 'cubic-bezier(0.0, 0.0, 0.2, 1)',
  acceleration: 'cubic-bezier(0.4, 0.0, 1, 1)',
  spring: 'cubic-bezier(0.175, 0.885, 0.320, 1.175)'
};

var ANIMATION_DURATION = 240;

var openAnimation = _uiBox.css.keyframes('openAnimation', {
  from: {
    transform: 'translateY(100%)'
  },
  to: {
    transform: 'translateY(0)'
  }
});

var closeAnimation = _uiBox.css.keyframes('closeAnimation', {
  from: {
    transform: 'scale(1)',
    opacity: 1
  },
  to: {
    transform: 'scale(0.9)',
    opacity: 0
  }
});

var animationStyles = {
  '&[data-state="entering"], &[data-state="entered"]': {
    animation: openAnimation + ' ' + ANIMATION_DURATION + 'ms ' + animationEasing.spring + ' both'
  },
  '&[data-state="exiting"]': {
    animation: closeAnimation + ' 120ms ' + animationEasing.acceleration + ' both'
  }
};

var CornerDialog = function (_PureComponent) {
  _inherits(CornerDialog, _PureComponent);

  function CornerDialog() {
    _classCallCheck(this, CornerDialog);

    var _this = _possibleConstructorReturn(this, (CornerDialog.__proto__ || Object.getPrototypeOf(CornerDialog)).call(this));

    _this.handleHidden = function () {
      _this.setState({ exiting: false });
      _this.props.onHide();

      if (_this.props.onExited) {
        var _this$props;

        (_this$props = _this.props).onExited.apply(_this$props, arguments);
      }
    };

    _this.handleClose = function () {
      _this.setState({ exiting: true });
      _this.props.onHide();
    };

    _this.state = {
      exiting: false
    };
    return _this;
  }

  _createClass(CornerDialog, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          width = _props.width,
          height = _props.height,
          hasCloseIcon = _props.hasCloseIcon,
          containerProps = _props.containerProps,
          isShown = _props.isShown,
          title = _props.title,
          onExit = _props.onExit,
          onExiting = _props.onExiting,
          onEnter = _props.onEnter,
          onEntering = _props.onEntering,
          onEntered = _props.onEntered;
      var exiting = this.state.exiting;


      return _react2.default.createElement(
        _portal.Portal,
        null,
        _react2.default.createElement(
          _Transition2.default,
          {
            appear: true,
            unmountOnExit: true,
            timeout: ANIMATION_DURATION,
            'in': isShown && !exiting,
            onExit: onExit,
            onExiting: onExiting,
            onExited: this.handleHidden,
            onEnter: onEnter,
            onEntering: onEntering,
            onEntered: onEntered
          },
          function (state) {
            return _react2.default.createElement(
              _layers.Pane,
              _extends({
                role: 'dialog',
                backgroundColor: 'white',
                elevation: 4,
                borderRadius: 8,
                width: width,
                height: height,
                css: animationStyles,
                'data-state': state,
                position: 'fixed',
                bottom: 8,
                right: 8
              }, containerProps),
              _react2.default.createElement(
                _layers.Pane,
                {
                  padding: 8,
                  borderBottom: 'extraMuted',
                  display: 'flex',
                  alignItems: 'center'
                },
                _react2.default.createElement(
                  _typography.Heading,
                  { is: 'h4', size: 500, flex: '1' },
                  title
                ),
                hasCloseIcon && _react2.default.createElement(_buttons.IconButton, {
                  height: 24,
                  icon: 'close',
                  appearance: 'ghost',
                  onClick: _this2.handleClose
                })
              ),
              _react2.default.createElement(
                _layers.Pane,
                { overflowY: 'auto', 'data-state': state, padding: 8 },
                typeof children === 'function' ? children({
                  close: close
                }) : children
              )
            );
          }
        )
      );
    }
  }]);

  return CornerDialog;
}(_react.PureComponent);

CornerDialog.propTypes = {
  title: _propTypes2.default.node,
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  height: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  isShown: _propTypes2.default.bool,
  children: _propTypes2.default.node,
  hasCloseIcon: _propTypes2.default.bool,
  containerProps: _propTypes2.default.object,

  onHide: _propTypes2.default.func,
  onExit: _propTypes2.default.func,
  onExiting: _propTypes2.default.func,
  onExited: _propTypes2.default.func,
  onEnter: _propTypes2.default.func,
  onEntering: _propTypes2.default.func,
  onEntered: _propTypes2.default.func
};
CornerDialog.defaultProps = {
  onHide: function onHide() {},
  width: 320,
  height: 176,
  hasCloseIcon: true
};
exports.default = CornerDialog;
