// evergreen-ui/esm/search-input/src/SearchInput.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Box, { splitBoxProps } from 'ui-box';
import { SearchIcon } from '../../icons';
import { getIconSizeForControlHeight } from '../../shared-styles';
import { TextInput } from '../../text-input';

var SearchInput = function (_PureComponent) {
  _inherits(SearchInput, _PureComponent);

  function SearchInput() {
    _classCallCheck(this, SearchInput);

    return _possibleConstructorReturn(this, (SearchInput.__proto__ || Object.getPrototypeOf(SearchInput)).apply(this, arguments));
  }

  _createClass(SearchInput, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          appearance = _props.appearance,
          iconProps = _props.iconProps,
          disabled = _props.disabled,
          height = _props.height,
          props = _objectWithoutProperties(_props, ['appearance', 'iconProps', 'disabled', 'height']);

      var _splitBoxProps = splitBoxProps(props),
          matchedProps = _splitBoxProps.matchedProps,
          remainingProps = _splitBoxProps.remainingProps;

      var _matchedProps$width = matchedProps.width,
          width = _matchedProps$width === undefined ? TextInput.defaultProps.width : _matchedProps$width;

      var iconSize = getIconSizeForControlHeight({ height: height });
      return React.createElement(
        Box,
        _extends({
          position: 'relative',
          display: 'inline-flex',
          height: height
        }, matchedProps),
        React.createElement(SearchIcon, _extends({
          pointerEvents: 'none',
          position: 'absolute',
          top: '50%',
          zIndex: 3,
          marginTop: -0.5 * height,
          iconSize: iconSize,
          size: height,
          disabled: disabled ? 'disabled' : 'default'
        }, iconProps)),
        React.createElement(TextInput, _extends({
          height: height,
          paddingLeft: height,
          appearance: appearance,
          disable: disabled,
          width: width
        }, remainingProps))
      );
    }
  }]);

  return SearchInput;
}(PureComponent);

SearchInput.propTypes = _extends({}, Box.propTypes, TextInput.propTypes, {
  iconProps: PropTypes.object
});
SearchInput.defaultProps = {
  height: 32,
  appearance: 'default'
};
export default SearchInput;
