// evergreen-ui/commonjs/search-input/src/SearchInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uiBox = require('ui-box');

var _uiBox2 = _interopRequireDefault(_uiBox);

var _icons = require('../../icons');

var _sharedStyles = require('../../shared-styles');

var _textInput = require('../../text-input');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SearchInput = function (_PureComponent) {
  _inherits(SearchInput, _PureComponent);

  function SearchInput() {
    _classCallCheck(this, SearchInput);

    return _possibleConstructorReturn(this, (SearchInput.__proto__ || Object.getPrototypeOf(SearchInput)).apply(this, arguments));
  }

  _createClass(SearchInput, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          appearance = _props.appearance,
          iconProps = _props.iconProps,
          disabled = _props.disabled,
          height = _props.height,
          props = _objectWithoutProperties(_props, ['appearance', 'iconProps', 'disabled', 'height']);

      var _splitBoxProps = (0, _uiBox.splitBoxProps)(props),
          matchedProps = _splitBoxProps.matchedProps,
          remainingProps = _splitBoxProps.remainingProps;

      var _matchedProps$width = matchedProps.width,
          width = _matchedProps$width === undefined ? _textInput.TextInput.defaultProps.width : _matchedProps$width;

      var iconSize = (0, _sharedStyles.getIconSizeForControlHeight)({ height: height });
      return _react2.default.createElement(
        _uiBox2.default,
        _extends({
          position: 'relative',
          display: 'inline-flex',
          height: height
        }, matchedProps),
        _react2.default.createElement(_icons.SearchIcon, _extends({
          pointerEvents: 'none',
          position: 'absolute',
          top: '50%',
          zIndex: 3,
          marginTop: -0.5 * height,
          iconSize: iconSize,
          size: height,
          disabled: disabled ? 'disabled' : 'default'
        }, iconProps)),
        _react2.default.createElement(_textInput.TextInput, _extends({
          height: height,
          paddingLeft: height,
          appearance: appearance,
          disable: disabled,
          width: width
        }, remainingProps))
      );
    }
  }]);

  return SearchInput;
}(_react.PureComponent);

SearchInput.propTypes = _extends({}, _uiBox2.default.propTypes, _textInput.TextInput.propTypes, {
  iconProps: _propTypes2.default.object
});
SearchInput.defaultProps = {
  height: 32,
  appearance: 'default'
};
exports.default = SearchInput;
