// evergreen-ui/commonjs/radio/src/RadioGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _layers = require('../../layers');

var _Radio = require('./Radio');

var _Radio2 = _interopRequireDefault(_Radio);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RadioGroup = function (_PureComponent) {
  _inherits(RadioGroup, _PureComponent);

  function RadioGroup(props, context) {
    _classCallCheck(this, RadioGroup);

    var _this = _possibleConstructorReturn(this, (RadioGroup.__proto__ || Object.getPrototypeOf(RadioGroup)).call(this, props, context));

    _this.handleChange = function (value) {
      _this.setState({
        selected: value
      });
      _this.props.onChange(value);
    };

    _this.state = {
      selected: (props.items.find(function (item) {
        return item.isDefaultSelected;
      }) || {}).value
    };
    return _this;
  }

  _createClass(RadioGroup, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          label = _props.label,
          items = _props.items,
          onChange = _props.onChange,
          props = _objectWithoutProperties(_props, ['label', 'items', 'onChange']);

      var selected = this.state.selected;


      return _react2.default.createElement(
        _layers.Pane,
        _extends({ role: 'group', 'aria-label': label }, props),
        items.map(function (item) {
          return _react2.default.createElement(_Radio2.default, {
            key: item.value,
            name: item.name,
            value: item.value,
            label: item.label,
            checked: selected === item.value,
            disabled: item.isDisabled,
            onChange: function onChange() {
              return _this2.handleChange(item.value);
            }
          });
        })
      );
    }
  }]);

  return RadioGroup;
}(_react.PureComponent);

RadioGroup.propTypes = _extends({}, _layers.Pane.propTypes, {
  label: _propTypes2.default.string,
  onChange: _propTypes2.default.func,
  items: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    name: _propTypes2.default.string,
    value: _propTypes2.default.string,
    label: _propTypes2.default.node,
    isDisabled: _propTypes2.default.bool,
    isDefaultSelected: _propTypes2.default.bool
  })).isRequired
});
RadioGroup.defaultProps = {
  items: []
};
exports.default = RadioGroup;
