// evergreen-ui/commonjs/radio/src/Radio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uiBox = require('ui-box');

var _uiBox2 = _interopRequireDefault(_uiBox);

var _typography = require('../../typography');

var _sharedStyles = require('../../shared-styles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var CircleIcon = function CircleIcon(_ref) {
  var size = _ref.size,
      _ref$fill = _ref.fill,
      fill = _ref$fill === undefined ? 'currentColor' : _ref$fill,
      props = _objectWithoutProperties(_ref, ['size', 'fill']);

  return _react2.default.createElement(
    'svg',
    _extends({ width: size, height: size, viewBox: '0 0 10 10' }, props),
    _react2.default.createElement('circle', { fill: fill, cx: '5', cy: '5', r: '5' })
  );
};

CircleIcon.propTypes = {
  fill: _propTypes2.default.string,
  size: _propTypes2.default.number
};

var Radio = function (_PureComponent) {
  _inherits(Radio, _PureComponent);

  function Radio() {
    _classCallCheck(this, Radio);

    return _possibleConstructorReturn(this, (Radio.__proto__ || Object.getPrototypeOf(Radio)).apply(this, arguments));
  }

  _createClass(Radio, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          id = _props.id,
          name = _props.name,
          label = _props.label,
          appearance = _props.appearance,
          disabled = _props.disabled,
          isInvalid = _props.isInvalid,
          checked = _props.checked,
          _onChange = _props.onChange,
          value = _props.value,
          size = _props.size,
          props = _objectWithoutProperties(_props, ['id', 'name', 'label', 'appearance', 'disabled', 'isInvalid', 'checked', 'onChange', 'value', 'size']);

      var appearanceStyle = _sharedStyles.CheckboxAppearances[appearance];

      return _react2.default.createElement(
        _uiBox2.default,
        _extends({
          is: 'label',
          cursor: disabled ? 'not-allowed' : 'pointer',
          display: 'flex',
          marginY: 12
        }, props),
        _react2.default.createElement(_uiBox2.default, _extends({
          is: 'input',
          id: id,
          type: 'radio',
          name: name,
          value: value,
          checked: checked,
          onChange: function onChange(e) {
            return _onChange(e.target.value);
          },
          disabled: disabled
        }, isInvalid ? { 'aria-invalid': true } : {}, {
          css: appearanceStyle
        })),
        _react2.default.createElement(
          _uiBox2.default,
          {
            boxSizing: 'border-box',
            borderRadius: 9999,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            marginTop: size === 12 ? 2 : 0,
            width: size,
            height: size
          },
          _react2.default.createElement(CircleIcon, { size: size === 12 ? 4 : 6 })
        ),
        label && _react2.default.createElement(
          _typography.Text,
          {
            marginLeft: 8,
            size: 300,
            color: disabled ? 'extraMuted' : 'default'
          },
          label
        )
      );
    }
  }]);

  return Radio;
}(_react.PureComponent);

Radio.propTypes = _extends({}, _uiBox2.default.propTypes, {
  appearance: _propTypes2.default.oneOf(Object.keys(_sharedStyles.CheckboxAppearances)),
  label: _propTypes2.default.node,
  disabled: _propTypes2.default.bool,
  isInvalid: _propTypes2.default.bool,
  checked: _propTypes2.default.bool,
  id: _propTypes2.default.string,
  name: _propTypes2.default.string,
  onChange: _propTypes2.default.func,
  value: _propTypes2.default.string,
  size: _propTypes2.default.oneOf([12, 16])
});
Radio.defaultProps = {
  appearance: 'default',
  onChange: function onChange() {},
  size: 12
};
exports.default = Radio;
